README.md
# iputils

R package containing utility functions to work with IP Addresses.

[![Travis-CI Build Status](https://travis-ci.org/bhaskarvk/iputils.png?branch=master)](https://travis-ci.org/bhaskarvk/iputils)

**NOTE** This is in pure R and will not be as fast as [iptools](https://www.github.com/hrbrmstr/iptools)

This is primarily for teaching myself (Bhaskar V. K.) how to write R packages.



CHANGELOG
2015-03-05  0.1
- Initial Commit
2015-03-06  0.1
- More progress towards releasing


tests/testthat.R
library(testthat)
library(iputils)

test_check("iputils")


tests/testthat/testipv4utils.R
library(iputils)

test_that('ipv4_to_long works correctly', {
  expect_identical(
    ipv4_to_long('127.0.0.1')[[1]],
    2130706433
  )
  expect_identical(
    ipv4_to_long(c('127.0.0.1','127.0.0.2')),
    list(2130706433,2130706434)
  )
})

test_that('par_ipv4_to_long works correctly', {
  expect_identical(
    par_ipv4_to_long('127.0.0.1')[[1]],
    2130706433
  )
  expect_identical(
    par_ipv4_to_long(c('127.0.0.1','127.0.0.2')),
    list(2130706433,2130706434)
  )
})

test_that('long_to_ipv4 works correctly', {
  expect_identical(
    long_to_ipv4(2130706433)[[1]],
    '127.0.0.1'
  )
  expect_identical(
    long_to_ipv4(c(2130706433,2130706434)),
    list('127.0.0.1','127.0.0.2')
  )
})

test_that('par_long_to_ipv4 works correctly', {
  expect_identical(
    par_long_to_ipv4(2130706433)[[1]],
    '127.0.0.1'
  )
  expect_identical(
    par_long_to_ipv4(c(2130706433,2130706434)),
    list('127.0.0.1','127.0.0.2')
  )
})

test_that('cidr_to_ipv4 works correctly', {
  expect_identical(
    cidr_to_ipv4('127.0.0.1/24')[[1]],
    as.list(paste('127.0.0',1:255,sep='.'))
  )
  expect_identical(
    cidr_to_ipv4(c('127.0.0.1/24','192.168.10.10/25')),
    list(
      as.list(paste('127.0.0',1:255,sep='.')),
      as.list(paste('192.168.10',10:127,sep='.'))
    )
  )
})

test_that('par_cidr_to_ipv4 works correctly', {
  expect_identical(
    par_cidr_to_ipv4('127.0.0.1/24')[[1]],
    as.list(paste('127.0.0',1:255,sep='.')))
  expect_identical(
    par_cidr_to_ipv4(c('127.0.0.1/24','192.168.10.10/25')),
    list(
      as.list(paste('127.0.0',1:255,sep='.')),
      as.list(paste('192.168.10',10:127,sep='.'))
    )
  )
})

test_that('cidr_to_range works correctly', {
  expect_identical(
    cidr_to_range('127.0.0.1/24')[[1]],
    '127.0.0.1-127.0.0.255'
  )
  expect_identical(
    cidr_to_range(c('127.0.0.1/24','192.168.10.10/25')),
    list(
      '127.0.0.1-127.0.0.255',
      '192.168.10.10-192.168.10.127'
    )
  )
})

test_that('par_cidr_to_range works correctly', {
  expect_identical(
    par_cidr_to_range('127.0.0.1/24')[[1]],
    '127.0.0.1-127.0.0.255'
  )
  expect_identical(
    par_cidr_to_range(c('127.0.0.1/24','192.168.10.10/25')),
    list(
      '127.0.0.1-127.0.0.255',
      '192.168.10.10-192.168.10.127'
    )
  )
})

test_that('range_to_ipv4 works correctly', {
  expect_identical(
    range_to_ipv4('127.0.0.1','127.0.0.255')[[1]],
    as.list(paste('127.0.0',1:255,sep='.'))
  )
  expect_identical(
    range_to_ipv4(
      c('127.0.0.1','192.168.10.10'),
      c('127.0.0.255','192.168.10.127')
    ),
    list(
      as.list(paste('127.0.0',1:255,sep='.')),
      as.list(paste('192.168.10',10:127,sep='.'))
    )
  )
})

test_that('par_range_to_ipv4 works correctly', {
  expect_identical(
    par_range_to_ipv4('127.0.0.1','127.0.0.255')[[1]],
    as.list(paste('127.0.0',1:255,sep='.'))
  )
  expect_identical(
    par_range_to_ipv4(
      c('127.0.0.1','192.168.10.10'),
      c('127.0.0.255','192.168.10.127')
    ),
    list(
      as.list(paste('127.0.0',1:255,sep='.')),
      as.list(paste('192.168.10',10:127,sep='.'))
    )
  )
})


tests/testthat/testipv6utils.R
library(iputils)

test_that('ipv4_to_mapped_ipv6 works correctly', {
  expect_identical(
    ipv4_to_mapped_ipv6('127.0.0.1')[[1]],
    '::FFFF:127.0.0.1'
  )
})

test_that('mapped_ipv6_to_ipv4 works correctly', {
  expect_identical(
    mapped_ipv6_to_ipv4('::FFFF:127.0.0.1')[[1]],
    '127.0.0.1'
  )
})



R/iputils.R
# Vectorized functions to convert ipv4 to/from integer
require(bitops) # required for unsigned int bitwise ops

int_to_ipv4 <- function(i) {
  lapply(i,function(x) {
    paste(bitAnd(bitShiftR(x,c(24,16,8,0)),0xFF),collapse='.')
    })
}

ipv4_to_int <- function(i) {
  lapply(i,function(x) {
    sum(bitShiftL(as.integer(strsplit(x,'.',fixed = T)[[1]]),c(24,16,8,0)))
  })
}

# Parallel versions
require(parallel)
par_int_to_ipv4 <- function(i) {
  mclapply(i,function(x) {
    paste(bitAnd(bitShiftR(x,c(24,16,8,0)),0xFF),collapse='.')
    })
}

par_ipv4_to_int <- function(i) {
  mclapply(i,function(x) {
    sum(bitShiftL(as.integer(strsplit(x,'.',fixed = T)[[1]]),c(24,16,8,0)))
  })
}


## the old_ are not as fast as ones w/o old_
old_cidr_to_ipv4 <- function(cidr) {
   sp <- strsplit(cidr,'/',fixed = T)
   numHosts <- lapply(sp,function(x) 2^(32-as.integer(x[[2]]))-1)
   start <- bitAnd(ipv4_to_int(lapply(sp,function(x) x[[1]])),
                   bitFlip(numHosts))+1
   mapply(function(s,n) {
     int_to_ipv4(seq(s ,length.out=n))
   } ,start, numHosts, SIMSIMPLIFY = F)
}

old_par_cidr_to_ipv4 <- function(cidr) {
   sp <- strsplit(cidr,'/',fixed = T)
   numHosts <- mclapply(sp,function(x) 2^(32-as.integer(x[[2]]))-1)
   start <- bitAnd(par_ipv4_to_int(mclapply(sp,function(x) x[[1]])),
                   bitFlip(numHosts))+1
   mcmapply(function(s,n) {
     int_to_ipv4(seq(s ,length.out=n))
   } ,start, numHosts, SIMPLIFY = F)
}


cidr_to_ipv4 <- function(cidr) {
   sp <- strsplit(cidr,'/',fixed = T)
   numHosts <- lapply(sp,function(x) 2^(32-as.integer(x[[2]]))-1)
   mapply(function(s,n) {
     int_to_ipv4(seq(bitAnd(ipv4_to_int(s)[[1]],bitFlip(n))+1
       ,length.out=n))
   } ,lapply(sp,function(x) x[[1]]),numHosts, SIMPLIFY = F)
}

par_cidr_to_ipv4 <- function(cidr) {
   sp <- strsplit(cidr,'/',fixed = T)
   numHosts <- mclapply(sp,function(x) 2^(32-as.integer(x[[2]]))-1)
   mcmapply(function(s,n) {
     par_int_to_ipv4(seq(bitAnd(ipv4_to_int(s)[[1]],bitFlip(n))+1
       ,length.out=n))
   } ,mclapply(sp,function(x) x[[1]]),numHosts, SIMPLIFY = F)
}

cidr_to_range <- function(cidr) {
   sp <- strsplit(cidr,'/',fixed = T)
   numHosts <- lapply(sp,function(x) 2^(32-as.integer(x[[2]]))-1)
   mapply(function(s,n) {
     start <- bitAnd(ipv4_to_int(s)[[1]],bitFlip(n))
     int_to_ipv4(c(start+1,start+n))
   } ,lapply(sp,function(x) x[[1]]),numHosts, SIMPLIFY = F)
}

par_cidr_to_range <- function(cidr) {
   sp <- strsplit(cidr,'/',fixed = T)
   numHosts <- mclapply(sp,function(x) 2^(32-as.integer(x[[2]]))-1)
   mcmapply(function(s,n) {
     start <- bitAnd(ipv4_to_int(s)[[1]],bitFlip(n))
     par_int_to_ipv4(c(start+1,start+n))
   } ,mclapply(sp,function(x) x[[1]]),numHosts, SIMPLIFY = F)
}

range_to_ipv4 <- function(start,end) {
  mapply(function(s,e){
    int_to_ipv4(seq(s,e))
  }, ipv4_to_int(start), ipv4_to_int(end), SIMPLIFY = F)
}

par_range_to_ipv4 <- function(start,end) {
  mcmapply(function(s,e){
    par_int_to_ipv4(seq(s,e))
  }, par_ipv4_to_int(start), par_ipv4_to_int(end), SIMPLIFY = F)
}


R/ipv6utils.R
#' Converts IPv4 to mapped IPv6
#' NAs will be mapped to NAs.
#'
#' @param v4 Vector of IP v4 Addresses
#' @return List of IP Addresses mapped IP v6 format
#' @examples
#' ipv4_to_mapped_ipv6('127.0.0.1')
#' ipv4_to_mapped_ipv6(c('127.0.0.1','192.168.1.1'))
ipv4_to_mapped_ipv6 <- function(v4) {
    lapply(v4, function(x) {
      if(is.na(x)) NA else
        paste('::FFFF',x,sep=':')
    })
}

#' Converts mapped IPv6 to IPv4
#' NAs will be mapped to NAs.
#'
#' @param v6 Vector of IP Addresses in mapped IPv6 format
#' @return List of IP Addresses in dotted notation IPv4 format
#' @examples
#' mapped_ipv6_to_ipv4('::FFFF:0:127.0.0.1')
#' mapped_ipv6_to_ipv4(c('::FFFF:0:127.0.0.1','::FFFF:0:192.168.1.1'))
mapped_ipv6_to_ipv4 <- function(v6) {
    lapply(v6, function(x) {
      if(is.na(x)) NA else
        tail(strsplit(x,':',fixed = T)[[1]],1)
    })
}


R/ipv4utils.R
#' Converts int to ipv4 dottend notation
#' NAs will be mapped to NAs.
#'
#' @param i Vector of IP Addresses in numeric format
#' @return List of IP Addresses in dotted notation format with length same as input
#' @examples
#' long_to_ipv4(2130706433)
#' long_to_ipv4(c(2130706433,2130706434))
long_to_ipv4 <- function(i) {
    lapply(i, function(x) {
      if(is.na(x)) NA else
        paste(bitops::bitAnd(bitops::bitShiftR(x, c(24, 16, 8, 0)), 255),
            collapse = ".")
    })
}

#' Converts ipv4 dottend notation to numeric
#' NAs will be mapped to NAs
#'
#' @param i Vector of IP Addresses in numeric format
#' @return List of IP Addresses in dotted notation format with length same as input
#' @examples
#' ipv4_to_long('127.0.0.1')
#' ipv4_to_long(c('127.0.0.1','192.168.1.1'))
ipv4_to_long <- function(i) {
    lapply(i, function(x) {
      if(is.na(x)) NA else
        sum(bitops::bitShiftL(as.integer(strsplit(x, ".", fixed = T)[[1]]),
            c(24, 16, 8, 0)))
    })
}

#' Converts int to ipv4 dottend notation in parallel
#' NAs will be mapped to NAs.
#'
#' @param i Vector of IP Addresses in numeric format
#' @return List of IP Addresses in dotted notation format with length same as input
#' @examples
#' par_long_to_ipv4(2130706433)
#' par_long_to_ipv4(c(2130706433,2130706434))
par_long_to_ipv4 <- function(i) {
    parallel::mclapply(i, function(x) {
      if(is.na(x)) NA else
        paste(bitops::bitAnd(bitops::bitShiftR(x, c(24, 16, 8, 0)), 255),
            collapse = ".")
    })
}

#' Converts ipv4 dottend notation to numeric in parallel
#'
#' @param i Vector of IP Addresses in numeric format
#' @return List of IP Addresses in dotted notation format with length same as input
#' @examples
#' par_ipv4_to_long('127.0.0.1')
#' par_ipv4_to_long(c('127.0.0.1','192.168.1.1'))
par_ipv4_to_long <- function(i) {
    parallel::mclapply(i, function(x) {
      if(is.na(x)) NA else
        sum(bitops::bitShiftL(as.integer(strsplit(x, ".", fixed = T)[[1]]),
            c(24, 16, 8, 0)))
    })
}

#' Converts IP v4 ranges in CIDR format to list of IP v4 addresses in dottend notation
#'
#' @param cidr Vector of IP Ranges in CIDR format
#' @return List of List of IP Addresses in dotted notation format with length same as input
#' @examples
#' cidr_to_ipv4('127.0.0.1/24')
#' cidr_to_ipv4(c('127.0.0.1/24','192.168.1.1/25'))
cidr_to_ipv4 <- function(cidr) {
  sp <- strsplit(cidr, "/", fixed = T)
  numHosts <- lapply(sp, function(x) 2^(32 - as.integer(x[[2]])))
  #start <- ipv4_to_long(lapply(sp,'[[',1))
  mapply(function(s, n) {
    end <- n - ( s - bitops::bitAnd(s,bitops::bitFlip(n-1)))
    long_to_ipv4(seq(s, length.out = end ))
  }, ipv4_to_long(lapply(sp, '[[', 1)), numHosts, SIMPLIFY = F)
}

#' Converts IP v4 ranges in CIDR format to list of IP v4 addresses in dottend notation in parallel
#'
#' @param cidr Vector of IP Ranges in CIDR format
#' @return List of List of IP Addresses in dotted notation format with length same as input
#' @examples
#' par_cidr_to_ipv4('127.0.0.1/24')
#' par_cidr_to_ipv4(c('127.0.0.1/24','192.168.1.1/25'))
par_cidr_to_ipv4 <- function(cidr) {
  sp <- strsplit(cidr, "/", fixed = T)
  numHosts <- parallel::mclapply(sp, function(x) 2^(32 - as.integer(x[[2]])))
  parallel::mcmapply(function(s, n) {
    end <- n - ( s - bitops::bitAnd(s,bitops::bitFlip(n-1)))
    par_long_to_ipv4(seq(s, length.out = end ))
  }, par_ipv4_to_long(parallel::mclapply(sp, '[[', 1)), numHosts, SIMPLIFY = F)
}

#' Converts IP v4 ranges in CIDR format to IP v4 ranges in dashed format
#'
#' @param cidr Vector of IP Ranges in CIDR format
#' @return List of IP v4 ranges in dashed format
#' @examples
#' cidr_to_ipv4('127.0.0.1/24')
#' cidr_to_ipv4(c('127.0.0.1/24','192.168.1.1/25'))
cidr_to_range <- function(cidr) {
  sp <- strsplit(cidr, "/", fixed = T)
  numHosts <- lapply(sp, function(x) 2^(32 - as.integer(x[[2]])))
  mapply(function(s, n) {
    end <- n - ( s - bitops::bitAnd(s,bitops::bitFlip(n-1)))
    paste(long_to_ipv4(c(s, s + end -1 )),collapse='-')
  }, ipv4_to_long(lapply(sp, '[[', 1)), numHosts, SIMPLIFY = F)
}

#' Converts IP v4 ranges in CIDR format to IP v4 ranges in dashed format in parallel
#'
#' @param cidr Vector of IP Ranges in CIDR format
#' @return List of IP v4 ranges in dashed format
#' @examples
#' par_cidr_to_ipv4('127.0.0.1/24')
#' par_cidr_to_ipv4(c('127.0.0.1/24','192.168.1.1/25'))
par_cidr_to_range <- function(cidr) {
  sp <- strsplit(cidr, "/", fixed = T)
  numHosts <- parallel::mclapply(sp, function(x) 2^(32 - as.integer(x[[2]])))
  parallel::mcmapply(function(s, n) {
    end <- n - ( s - bitops::bitAnd(s,bitops::bitFlip(n-1)))
    paste(par_long_to_ipv4(c(s, s + end -1)),collapse='-')
  }, par_ipv4_to_long(parallel::mclapply(sp, '[[', 1)), numHosts, SIMPLIFY = F)
}

#' Converts IP v4 ranges in CIDR format to IP v4 Addresses
#'
#' @param start Vector of Starting IP v4 address in dotted notation
#' @param end Vector of Starting IP v4 address in dotted notation
#' @return List of List of IP Addresses in dotted notation format with length same as input
#' @examples
#' range_to_ipv4('127.0.0.1','127.0.0.100')
#' range_to_ipv4(c('127.0.0.1','192.168.1.1'), c('127.0.0.100','192.168.1.255'))
range_to_ipv4 <- function(start, end) {
    mapply(function(s, e) {
        long_to_ipv4(seq(s, e))
    }, ipv4_to_long(start), ipv4_to_long(end), SIMPLIFY = F)
}

#' Converts IP v4 ranges in CIDR format to IP v4 Addresses in parallel
#'
#' @param start Vector of Starting IP v4 address in dotted notation
#' @param end Vector of Starting IP v4 address in dotted notation
#' @return List of List of IP Addresses in dotted notation format with length same as input
#' @examples
#' par_range_to_ipv4('127.0.0.1','127.0.0.100')
#' par_range_to_ipv4(c('127.0.0.1','192.168.1.1'), c('127.0.0.100','192.168.1.255'))
par_range_to_ipv4 <- function(start, end) {
    parallel::mcmapply(function(s, e) {
        par_long_to_ipv4(seq(s, e))
    }, par_ipv4_to_long(start), par_ipv4_to_long(end),
        SIMPLIFY = F)
}
